### ISSUE TITLE
Keybindings aren't respecting `when clause` when showing shortcut keys in menus

### ISSUE THREAD

**Issue by eamodio** [Microsoft/vscode#40389]
- VSCode Version: Code - Insiders 1.20.0-insider (08ec16001adaab8a86183431f6aec09209caa4ab, 2017-12-15T05:15:38.886Z)
- OS Version: Windows_NT x64 10.0.17046
---

Originally reported https://github.com/eamodio/vscode-gitlens/issues/229

>Steps to Reproduce:
>
>1. I set it up "gitlens.keymap": "none" or "gitlens.keymap": "standard"
>2. The shortcut menu of the editor menu does not change.
![1](https://user-images.githubusercontent.com/4232281/33921773-13f8cce8-e001-11e7-94cc-d275ed810d15.png)
![2](https://user-images.githubusercontent.com/4232281/33921775-170c70c4-e001-11e7-9fef-51ad9b77faa1.png)

For context, I've recently added a new setting (and associated command context) to control the set of keybindings GitLens adds. It seems that when menu items for a command that also has a keybinding, doesn't honor the `when` clause of the keybinding (for showing the shortcut in the menu). Instead it seems to just grab the last keybinding found for that command in the `package.json`

Reproduces without extensions: No

**Comment by usernamehw** [comment]
A duplicate of https://github.com/Microsoft/vscode/issues/35225

**Comment by eamodio** [comment]
I don't think that is a duplicate -- I looked into that issue before I posted, but while it is somewhat related (when clause on keybindings), the issue I'm reporting is purely display only. It is showing the wrong shortcut key in the menu -- everything else works properly.

**Comment by ramya-rao-a** [comment]
@eamodio 

This is similar to the concern I raised in https://github.com/Microsoft/vscode/issues/27611#issuecomment-305113399

Look at the explanation from @alexandrudima in https://github.com/Microsoft/vscode/issues/27611#issuecomment-307008730

**Comment by eamodio** [comment]
@ramya-rao-a thanks! That sounds very similar.

This is definitely a tricky one. This might be a bit hacky, but what if there was a new option on a `keybinding` something like `shownWhen` (or `menuWhen`) -- which if it is missing, is always true (so the current behavior stays). But an extension can provide a specific when clause that **should** be checked before showing the shortcuts in the menus?

Then I could have something like:
```json
...
{
    "command": "gitlens.toggleCodeLens",
    "key": "shift+alt+b",
    "shownWhen": "gitlens:keymap == standard",
    "when": "gitlens:keymap == standard && editorTextFocus && gitlens:activeIsTracked && gitlens:canToggleCodeLens"
},
{
    "command": "gitlens.toggleCodeLens",
    "key": "ctrl+alt+g shift+b",
    "mac": "cmd+alt+g shift+b",
    "shownWhen": "gitlens:keymap == chorded",
    "when": "gitlens:keymap == chorded && editorTextFocus && gitlens:activeIsTracked && gitlens:canToggleCodeLens"
},
...
```

Probably easiest to just have the `when` clause duplicate the `shownWhen` as to not conflate the two.

Another nice "feature" is, if for some reason an extension wanted to hide a shortcut key from a menu, it could `"showWhen": false` or something like that.

Thoughts?

**Comment by alexandrudima** [comment]
This is tough.

Our keybinding system is a rule-based system, which, given a certain concrete context, will resolve a certain keypress to a command. If we were to give that a notation, it would be `f(context,keybinding) => command`

Today, displaying of keybindings next to a command title (what the menus do) is already tricky because it accounts for *statical shadowing*, but it does not account for a specific context. If we were to give that a notation, it would be `f-1(command)=>keybinding`. 

`f-1` is implemented in code as [`IKeybindingService.lookupKeybinding`](https://github.com/Microsoft/vscode/blob/bcc46316b497d1a316dd47ae86481df3ca5784ee/src/vs/platform/keybinding/common/keybinding.ts#L71). In this specific case, it finds both rules, and because it doesn't know which one really applies, it will pick the *last written one*.

IMHO it would be *costly* to begin passing in a context to that method. i.e. to change `f-1` to also accept a certain context as an argument. The reason is that the context is something highly dynamic. So, on every click, on every selection change, etc, the context *might* change, which means the results of `f-1` *might* change, which means we need to repaint possibly hundreds of keybindings labels because they *might* change. We *might* need to push messages to the main process to update the top-level menus too.

**Comment by eamodio** [comment]
So 2 more thoughts on this. I noticed when hooking this up in GitLens, that I was unable to use `config.*` in the `when` clause for a keybinding (which is different than commands). But if we were able to use `config.*` in the keybinding `when`, then maybe they could be used for the case where you have to choose between showing more than 1 keybinding, since configuration is in theory far more stable and less contextual?

Or another less ideal option, would be to maybe just not show any keybinding if it can't be determined which one should be shown? Is there known cases where showing the last written one is correct (other than being deterministic)?

**Comment by eamodio** [comment]
This is causing more and more confusion for my GitLens users -- is there something I can do here to avoid this?

@alexandrudima instead of using the context always, could it just be used as a disambiguator for when one definitive binding can't be found?

**Comment by alexandrudima** [comment]
@eamodio Sorry about that.

The only workarounds on your side that I can think of are to:
* not ship with all those keybindings
* (and) guide the end user towards copy-pasting the keybindings rules in their `keybindings.json`.
* (or) creating separate gitlens keymap 1, gitlens keymap 2 extensions which would contribute only the keybindings.

**Comment by alirezaghey** [comment]
Hi @alexandrudima 
I was wondering if this issue could be opened again? It's weird and causes a lot of trouble to get to the bottom of it.

**Comment by eamodio** [comment]
@alexdima It's been a while, and this is still an issue affecting users. Should this just get closed out-of-scope? Or is there anything else that can be done here? Maybe only pick the _last written keybinding, without a when clause_ and if there are none without a _when_ clause, don't show a shortcut key at all? IMO, have no shortcut shown is better than the wrong one.

**Comment by alexdima** [comment]
I think two distinct things could be done:
- the lookup keybindings code could become aware of `config.*` context keys, since it is safe to trust that the `config.*` values are not changing all the time (for example on focus change). I looked into this briefly, but it is not trivial. Then, the gitlens extension could define keybindings and switch keymaps using a `config.*` context key.
- the other idea is to ship one or more gitlens keymap extensions separately from the main gitlens extension. IMHO it is not good to bundle great functionality with keybindings, since keybindings are so personal and many people want the functionality, but not the keybindings. So there could be gitlens-keymap1 and gitlens-keymap2 and users could install 0, 1, or 2 of those, as they wish.

**Comment by eamodio** [comment]
GitLens keys are currently conditional based on a `config.*` setting.

I thought about the keymap option, but I'm definitely worried about migration and a bit with discovery. I've also thought about not having any keybindings or keymaps. And completely leaving it up to users -- and for migration, if users want to keep their current keybindings, I could just write them out into the `keybindings.json`. And I think that would probably work decently well (especially since we now allow you to more easily set keybindings right from the command palette.

The biggest thing stopping me from that (besides just the work), is for certainly GitLens commands (_Open Changes with Previous Revision_, _Open Changes with Next Revision_, etc) are the complex `when` conditions that should accompany them.

For example, this is what you'd really need to add to customize the _Open Changes with Previous Revision_ command:

```json
{
	"command": "gitlens.diffWithPrevious",
	"key": "alt+,",
	"when": "editorTextFocus && gitlens:activeFileStatus =~ /tracked/ && !isInDiffEditor"
},
{
	"command": "gitlens.diffWithPreviousInDiffLeft",
	"key": "alt+,",
	"when": "editorTextFocus && gitlens:activeFileStatus =~ /tracked/ && isInDiffEditor && !isInDiffRightEditor"
},
{
	"command": "gitlens.diffWithPreviousInDiffRight",
	"key": "alt+,",
	"when": "editorTextFocus && gitlens:activeFileStatus =~ /tracked/ && isInDiffRightEditor"
},
```
